compiler/rustc_arena/src/lib.rs

```diff
@@ -282,7 +304,89 @@ impl<T> TypedArena<T> {
     }
 }
 
-unsafe impl<#[may_dangle] T> Drop for TypedArena<T> {
+impl<T> TypedArena<T> {
+    /// Allocates an object in the `TypedArena`, returning a reference to it.
+    #[inline]
+    pub fn alloc(&self, object: T) -> &mut T {
+        self.scalar.alloc(object)
+    }
+
+    /// Allocates a slice of objects that are copied into the `TypedArena`, returning a mutable
+    /// reference to it. Will panic if passed a zero-sized types.
+    ///
+    /// Panics:
+    ///
+    ///  - Zero-sized types
+    ///  - Zero-length slices
+    #[inline]
+    pub fn alloc_slice(&self, slice: &[T]) -> &mut [T]
+    where
+        T: Copy,
+    {
+        self.scalar.alloc_slice(slice)
+    }
+
+    #[inline]
+    pub fn alloc_from_iter_general<I: Iterator<Item = T>>(&self, iter: I) -> &mut [T] {
```

Comment:
can this now be invoked with zero sized types? Or should this function be private?

Reply:
It should be private